runtime/src/chpl-cache.c

```diff
@@ -1673,69 +1855,148 @@ void validate_cache(struct rdcache_s* tree)
-    if( at <= sn ) {
+    if (index == -1) break;
+
+    // If the first entry's sequence number is earlier than sn
+    // and that comm event isn't complete yet, then wait for some
+    if (cache->pending_sequence_numbers[index] <= sn &&
+        !chpl_comm_test_nb_complete(cache->pending[index])) {
       // Wait for some requests
       last = cache->pending_last_entry;
-      if( last < index ) last = cache->pending_len - 1;
-      while( cache->pending[index] ) {
-        DEBUG_PRINT(("wait_for waiting %i..%i\n", index, last));
-        // Wait for some requests to complete.
-        chpl_comm_wait_nb_some(&cache->pending[index], last - index + 1);
+      if (last < index) last = cache->pending_len - 1;
+
+      // Wait for some requests to complete.
+      // (this could cause a different task body to run)
+      chpl_comm_wait_nb_some(&cache->pending[index], last - index + 1);
+
+      if (EXTRA_YIELDS) {
+        TRACE_YIELD_PRINT(("%d: task %d cache %p yielding in do_wait_for "
+                           "for chpl_comm_wait_nb_some\n",
+                           chpl_nodeID, (int) chpl_task_getId(), cache));
+
+        chpl_task_yield();
+
+        TRACE_YIELD_PRINT(("%d: task %d cache %p back in do_wait_for\n",
+                          chpl_nodeID, (int) chpl_task_getId(), cache));
       }
-      // cache->pending[index] == NULL now
+
+      // continue the loop
+    }
+
+    // Whether we waited above or not, if the first entry's event
+    // is already complete, then remove it from the queue.
+    if (chpl_comm_test_nb_complete(cache->pending[index])) {
+      fifo_circleb_pop(&cache->pending_first_entry,
+                       &cache->pending_last_entry,
+                       cache->pending_len);
+      cache->completed_request_number =
+        seqn_max(cache->completed_request_number,
+                 cache->pending_sequence_numbers[index]);
+
+      TRACE_NB_PRINT(("%d: task %d wait_for popped %i completed=%i\n",
+                      chpl_nodeID, (int)chpl_task_getId(),
+                      (int) cache->pending_sequence_numbers[index],
+                      (int) cache->completed_request_number));
     }
-    if( chpl_comm_test_nb_complete(cache->pending[index]) ) {
-      // we completed cache->pending[index], so remove the entry from the queue.
-      DEBUG_PRINT(("wait_for removing %i\n", index));
-      fifo_circleb_pop( &cache->pending_first_entry, &cache->pending_last_entry, cache->pending_len);
-      max_completed = seqn_max(max_completed, at);
-    } else if( at > sn ) {
-      // Stop if we have an uncompleted request for a later sequence number
+
+    // Stop if we have an uncompleted request for a later sequence number
+    if (cache->pending_sequence_numbers[index] > sn) {
       DEBUG_PRINT(("wait_for stopped at %i\n", index));
       break;
     }
   }
+}
+
+
+// "lock" the entry
+//  * if entry->entryReservedByTask is task_local, return - already locked
```

Comment:
Has this comment gone out of date, or perhaps actually describe some other code (the caller?) rather than this function itself? It definitely doesn't match the logic in the function.

Reply:
Yes, it's gone out of date. Thanks for noticing. I've just pushed a fix to the comment.